stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -0,0 +1,127 @@
+ * <code>min_block</code> then the cholesky decomposition on the GPU is computed
+ * using that submatrix. If <code>block</code> is greater than
+ * <code>block_size</code> then <code>cholesky_decompose</code> is run again
+ * with <code>block</code> equal to <code>block/divider</code>. Once the
+ * Cholesky Decomposition is computed, the full matrix cholesky is created
+ * by propogating the cholesky forward as given in the reference report below.
+ *
+ * For a full guide to how this works
+ * see the Cholesy decompostion chapter in the  reference report
+ * <a href="https://goo.gl/6kWkJ5"> here</a>.
+ * @param A Symmetric matrix on the GPU.
+ * @param block Size of the block used to compute the cholesky decomposition.
+ * @param divider Proportion to divide the submatrix by at each recursive step.
+ * @param min_block The amount that block is checked against to decide
+ *  whether to continue the recursion or to perform the cholesky.
+ * @return Square root of matrix on the GPU.
+ * @throw std::domain_error if m is not
+ *  positive definite (if m has more than 0 elements)
+ */
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block,
+                                     const int divider, const int min_block) {
+  auto offset = 0;
+  // NOTE: The code in this section follows the naming conventions
+  // in the report linked in the docs.
+  matrix_gpu A_11(block, block);
+  // Repeats the blocked cholesky decomposition until the size of the remaining
+  // submatrix is smaller or equal to the block size
+  while ((offset + block) < (A.rows())) {
+    auto block_subset = A.rows() - offset - block;
+    matrix_gpu A_21(block_subset, block);
+    matrix_gpu A_22(block_subset, block_subset);
+    // Copies a block of the input A into A_11
+    A_11.sub_block(A, offset, offset, 0, 0, block, block);
+    // The following function either calls the
+    // blocked cholesky recursively for the submatrix A_11
+    // or calls the kernel  directly if the size of the block is small enough
+    matrix_gpu L_11 = stan::math::internal::cholesky_decompose_recursion(
+        A_11, block, divider, min_block);
+    // Copies L_11 back to the input matrix
+    A.sub_block(L_11, 0, 0, offset, offset, block, block);
+    // Copies a block of the input A into A_21
+    auto block_offset = offset + block;
+    A_21.sub_block(A, block_offset, offset, 0, 0, block_subset, block);
+    // computes A_21*((L_11^-1)^T)
+    // and copies the resulting submatrix to the input matrix
+    auto A_11_inverse_T = transpose(lower_triangular_inverse(L_11));
+    matrix_gpu L_21 = A_21 * A_11_inverse_T;
+    A.sub_block(L_21, 0, 0, block_offset, offset, block_subset, block);
+    A_22.sub_block(A, block_offset, block_offset, 0, 0, block_subset,
+                   block_subset);
+    // computes A_22 - L_21*(L_21^T)
+    matrix_gpu temp = multiply_transpose(L_21);
+    matrix_gpu L_22 = A_22 - temp;
```

Comment:
Sweet. We don't have to hold this PR up for it, but do you think you could make an issue for this idea? I'd try but I think it would be too vague as I don't totally know what it takes.